### PR TITLE
Feature - CLI Tutor Mode

### PR DESCRIPTION
**Pull request by jay-dee7**
Adds the **CLI Tutor Mode**, which can help with learning the IPFS Command Line Interface from the Webui/Desktop. This feature was requested in #1421 

#### Done
* Option to enable/disable CLI Tutor Mode from the settings page
* If enabled, a `copy-icon` 📋 is shown next to the buttons/menus where those actions can be performed via IPFS' CLI
* This icon is visible at:
  * Files Page -> inside `+ Import` menu (Add File, Folder, Copy from IFPS)
  * Files Page -> File Menu (three dots next to a file) -> (Delete, Rename, Pin/Unpin, Download)
  * Peers Page -> next to `+ Add Connection`
  * Settings Page -> next to `Reset` button in config edit box
  * ~~Setting Page -> next to `Submit` button in API edit box~~

### Pending Work:
* ~~I recently discovered a state management issue with the modals, I couldn't find the root cause so far, it started happening after the refactor today (1st Aug 2020)~~
* ~~After rebasing from master today, there's a small UI glitch (text in the modal actions is left aligned, I'll work it out before making this PR reviewable)~~
* ~~I have added a bunch of TODO's in the code, which can potentially help with the code cleanup~~
* ~~Files component kind of looks like a mess right now, a lot of state is being shared across mulitple components (maybe we can refactor it to make it look neat)~~

Here's a visual for how it looks in action: [Jessica note: these are a bit outdated now since commits for visual tweaks on 12 Aug]

### Settings Page | Option to enable CLI Tutor Mode
enable-ipfs-cli-tutor-mode

### Files Page | CLI Tutor Mode Enabled
cli-tutor-mode-files-page

### Settings Page | CLI Tutor Mode - Open Modal
cli-tutor-mode-open-modal

Signed-off-by: Jasdeep Singh <jasdeepsingh.uppal@gmail.com>

### CODE REVIEW

**Comment by jay-dee7** [comment]
@jessicaschilling here's the draft for cli tutor mode, the place which would need more attention is files component, since it has a lot of shared state, please check this out and suggest any refactors/changes for places where i might have broken any acceptable coding patterns. I'll update this PR tomorrow with little of code cleanup and modal actions alignment fix

**Comment by jessicaschilling** [comment]
Thanks, @jay-dee7! I'd like @rafaelramalho19 to look through this in more detail ... but just a heads up that he's on holiday for the next week. Don't want you to think that we are ignoring you. 😊

**Comment by jay-dee7** [comment]
That's okay @jessicaschilling. In-fact apologies for the delay. This should've been up way before than it did. In the meantime I'll keep  on iterating  over this, just to see if anything can be improved upon 😄

File: src/components/cli-tutor-mode/CliTutorMode.js
Comment (jessicaschilling) [inline src/components/cli-tutor-mode/CliTutorMode.js:21]:
```suggestion
        <p className='charcoal w-80 center'>{t('Copy the following and paste it into your terminal application to do this task in IPFS via the command line.')}</p>
```
Just a text edit.

File: src/components/cli-tutor-mode/CliTutorMode.js
Comment (jessicaschilling) [inline src/components/cli-tutor-mode/CliTutorMode.js:23]:
```suggestion
          <Shell title="Shell">
```
Just "Shell" is fine here. Use the capital "S" in case we decide not to use CSS to uppercase these lines in future.

**Comment by jay-dee7** [comment]
sure @jessicaschilling i'll make these changes ASAP. There are more things that I didn't notice initially. ContextMenu in files pages needs some work, after rebasing, the copy icon and the text aren't aligned symmetrically. I'll try to get them fixed before the review starts

**Comment by jay-dee7** [comment]
@jessicaschilling I rebased from master just a while ago, one of the incoming changes were in `ApiAddressForm` component and since it was written using hooks (I believe), I was not able to overwrite props easily (I'm not sure what exactly broke the prop sharing, i believe its hooks but i have no idea.). So just for the `ApiAddressForm` component, I've managed the state for `cli-tutor-modal` locally to the component, instead of using the regular redux actions. Rest of the stuff stays almost same. I think this PR is ready for review 🤞

**Comment by jessicaschilling** [comment]
Thanks, @jay-dee7! 
@rafaelramalho19 - are you able to take a look at this PR once you are back in and settled? Thank you!

File: src/components/cli-tutor-mode/CliTutorMode.js
Comment (rafaelramalho19) [inline src/components/cli-tutor-mode/CliTutorMode.js:56]: You're using a functional component above and here you used a class component. Can you make them equal? (I would advise choosing functional)

File: src/components/cli-tutor-mode/CliTutorMode.js
Comment (rafaelramalho19) [inline src/components/cli-tutor-mode/CliTutorMode.js:25]: Please change this to `command === cliCmdKeys.UPDATE_IPFS_CONFIG` (you need to import cliCmdKeys of course).

We should have a single source of truth for the command, just to avoid changing it on the constant files and breaking on this file.

File: src/components/cli-tutor-mode/CliTutorMode.js
Comment (rafaelramalho19) [inline src/components/cli-tutor-mode/CliTutorMode.js:40]: Replace here as suggested previously.

Also, why do you check for the existence of `command` here but not on line 25?

File: src/components/cli-tutor-mode/CliTutorMode.js
Comment (rafaelramalho19) [inline src/components/cli-tutor-mode/CliTutorMode.js:85]: Just to be consistent shouldn't you do `() => doOpenCliTutorModal(false)`? I think that's the issue you had with being an object sometimes (I think that onLeave sends the click event as an argument)

File: src/components/cli-tutor-mode/CliTutorMode.js
Comment (rafaelramalho19) [inline src/components/cli-tutor-mode/CliTutorMode.js:98]: You're not using tReady, why declare it?

File: src/files/FilesPage.js
Comment (rafaelramalho19) [inline src/files/FilesPage.js:240]: Todos are actually done with `//TODO: `, and isn't it connected already?

File: src/files/dropdown/Dropdown.js
Comment (rafaelramalho19) [inline src/files/dropdown/Dropdown.js:13]: Try to avoid the transform (same as commented before)

File: src/settings/SettingsPage.js
Comment (rafaelramalho19) [inline src/settings/SettingsPage.js:67]: You need to add a translation key here, not the text

File: src/settings/SettingsPage.js
Comment (rafaelramalho19) [inline src/settings/SettingsPage.js:70]: Same here.

File: src/settings/SettingsPage.js
Comment (rafaelramalho19) [inline src/settings/SettingsPage.js:76]: And here

File: src/settings/SettingsPage.js
Comment (lidel) [reply src/settings/SettingsPage.js:67]: Namely, add the key to `public/locales/en/settings.json` 
(no need to add it to other locales, that will happen automatically after PR is merged)

File: src/bundles/cli-tutor-mode.js
Comment (jay-dee7) [reply src/bundles/cli-tutor-mode.js]: @rafaelramalho19 this was bad on my part, this piece of code isn't required at all, I cleaned it up in my latest commit

File: src/bundles/cli-tutor-mode.js
Comment (jay-dee7) [reply src/bundles/cli-tutor-mode.js]: @rafaelramalho19 I did add the suggested pattern however this was the first time I used redux-bundler. Please go through it once more just to validate if I did what was requested. I hope i didn't drift away too far :)

File: src/components/api-address-form/ApiAddressForm.js
Comment (jay-dee7) [reply src/components/api-address-form/ApiAddressForm.js]: for this, i have now added two options, one using just the bigger pixel size (height is increased) and another one using `msTransform` and `webkitTransform`. I can quickly stick to one solution but as far as i can tell, they both work as expected. `webkitTransform` worked okay for me on safari too

File: src/components/cli-tutor-mode/CliTutorMode.js
Comment (jay-dee7) [reply src/components/cli-tutor-mode/CliTutorMode.js]: I've Converted this into a functional component :)

File: src/components/cli-tutor-mode/CliTutorMode.js
Comment (jay-dee7) [reply src/components/cli-tutor-mode/CliTutorMode.js]: Thanks for digging this one, I've added comparison with the full command from `cliCmdKeys` now

File: src/components/cli-tutor-mode/CliTutorMode.js
Comment (jay-dee7) [reply src/components/cli-tutor-mode/CliTutorMode.js]: added the same comparison here too

File: src/components/cli-tutor-mode/CliTutorMode.js
Comment (jay-dee7) [reply src/components/cli-tutor-mode/CliTutorMode.js]: I think thats exactly what was causing the issue. I replaced the call with `() => doOpenCliTutorModal(false/true)` and it has been working just fine

File: src/components/cli-tutor-mode/CliTutorMode.js
Comment (jay-dee7) [reply src/components/cli-tutor-mode/CliTutorMode.js]: Removed this and added `command`, since command is a required field

File: src/files/FilesPage.js
Comment (jay-dee7) [reply src/files/FilesPage.js]: contextMenu page actually isn't connected but also it doesn't need to. Maybe if we'd want, we do can connect `Dropdown` component to reduce prop chaining

File: src/files/dropdown/Dropdown.js
Comment (jay-dee7) [reply src/files/dropdown/Dropdown.js]: added a css class for this with `webkitTransform` but it would also work if we simple increased the height by 1.3X. Lemme know which way we'd like to move forward with :)

File: src/settings/SettingsPage.js
Comment (jay-dee7) [reply src/settings/SettingsPage.js]: @rafaelramalho19 @lidel i've replace these text lines with keys from translation files :)

File: src/settings/SettingsPage.js
Comment (jay-dee7) [reply src/settings/SettingsPage.js]: replaced it

File: src/settings/SettingsPage.js
Comment (jay-dee7) [reply src/settings/SettingsPage.js]: and replaced this one too

File: src/components/api-address-form/ApiAddressForm.js
Comment (rafaelramalho19) [reply src/components/api-address-form/ApiAddressForm.js]: It works, the problem is that it renders a pixelated icon in safari (old issue unfortunately), so I would just stick with increasing the height

File: src/bundles/cli-tutor-mode.js
Comment (rafaelramalho19) [reply src/bundles/cli-tutor-mode.js]: You did good 👍

**Comment by rafaelramalho19** [comment]
After you remove the transforms, LGTM 😄

**Comment by jay-dee7** [comment]
thanks for the amazing review @rafaelramalho19, i've change the css class to use just height and no transforms :)

**Comment by jessicaschilling** [comment]
@lidel -- do you mind going over this when you're able? I think we're very close to this being done.

NOTE: I want to get the "copy to clipboard" icon into ipfs-css before we ship. I can do that while you're reviewing 😊

Thanks!

**Comment by jay-dee7** [comment]
hey @jessicaschilling I just saw the changes that you made and I think they really addch the finishing touches to the feature, left aligned text in the modal looks great plus the new icon actually looks pretty amazing :) However, I did notice once thing though, the Checkbox on CLI Tutor Mode is slightly moved to bottom `lh-copy` class is doing the most of it. Can we do some sort of workaround for that?

I was going through the ipfs documentation and realised that a couple of commands are not correct I think, namely
* `Files Page -> + Import Button -> New Folder`
* `Files Page -> individual file -> three dots -> Rename`
I'll push the change for both of these commands just now but please let me know if rest of them are correct too

image

File: src/bundles/files/consts.js
Comment (lidel) [inline src/bundles/files/consts.js:66]:
```suggestion
  [cliCmdKeys.CREATE_NEW_DIRECTORY]: () => 'ipfs files mkdir <folder-name>',
```

**Comment by jessicaschilling** [comment]
@jay-dee7 @lidel I'd vote for closing the dialog onclick of the "Copy" button. @jay-dee7, can you please change that?

**Comment by jay-dee7** [comment]
Sure @jessicaschilling I'll make the change ASAP :)

**Comment by jessicaschilling** [comment]
@jay-dee7 And thanks for spotting the misaligned checkbox! Fixed it. 😊

**Comment by jay-dee7** [comment]
@jessicaschilling @lidel i've added both the requested points. 
Now by clicking the copy button in the modal, we copy the contents (command) and then close the modal.

Also, I've added a fix for rename command. I realized that we need to preserve the file path while renaming and we should wrap the full path path with quotes ("<path>"), because it'll throw errors if the full path contains spaces (for example a file named `cat image.png` would throw error in cli but if we wrap it with quotes as `"cat image.png"`) it won't. Please let me know if we're good to go with this :)

**Comment by jessicaschilling** [comment]
@lidel - I'm happy but will wait for your OK, please feel free to merge 😊

File: tsconfig.json
Comment (lidel) [inline tsconfig.json:7]: @Gozala how strongly do we  feel about this?
@jay-dee7 is it possible to add type info and keep this as `true`?
```suggestion
    "noImplicitAny": true,
```

File: src/files/modals/Modals.js
Comment (lidel) [inline src/files/modals/Modals.js:163]: I believe WebUI pins recursively:
```suggestion
        return cliCommandList[action](activeCid, isPinned ? 'rm' : 'add -r')
```

**Review by lidel — CHANGES_REQUESTED** [review]
@jay-dee7 nearly there! small asks inline :pray:

File: src/files/modals/Modals.js
Comment (jay-dee7) [reply src/files/modals/Modals.js:163]: @lidel by recursive you mean that if we pin a directory, all the files in the directory and any of it's sub-directories and their files gets pinned?

File: src/bundles/files/consts.js
Comment (jay-dee7) [reply src/bundles/files/consts.js:60]: @lidel thanks for catching this one. I'll change it to use full filepath and also recursive. I think 
```ipfs files rm -r ${filepath}```
is a much safer option (command would succeed in more cases regardless of a directory/file)

File: src/files/modals/Modals.js
Comment (jay-dee7) [reply src/files/modals/Modals.js:163]: i just confirmed, webui doesn;t add/remove pin recursively
ipfs-add-pin
 but please let me know which approach should i follow :)

File: tsconfig.json
Comment (jay-dee7) [reply tsconfig.json]: I just added type information for the methods where typescript transpiler was failing and reverted this option back to `true`. It works now as expected :)

**Review by lidel — APPROVED** [review]
Thank you @jessicaschilling, I think the old i18next issue is gone, all looks good.

----

I've made a small fix in 5641184 to ensure CLI tutor lists import actions as the button in GUI 
(`ipfs add && ipfs files cp`, context: https://github.com/ipfs-shipyard/ipfs-webui/issues/676)

Ideally `/<dest-name>` in import commands would be replaced with `"${path}/<dest-name>"` where `path` is the current dir in MFS. I started doing that in 3a337de but had no bandwidth to wire it up.

I think it's ok to merge with static destination, but if anyone has time to ensure `path` is set for [these actions](https://github.com/jay-dee7/ipfs-webui/blob/jay-dee7/cli-tutor-mode/src/files/modals/Modals.js#L157-L160) that would be even better for user experience.

**Comment by jessicaschilling** [comment]
@lidel -- Thanks for calling the subdirectory detail out. I'm inclined to merge this as-is and address that in a separate issue (https://github.com/ipfs-shipyard/ipfs-webui/issues/1601), since I suspect we may want to make some tweaks to tutor mode overall after this gets into the hands of users for a bit.

I've also opened https://github.com/ipfs-shipyard/ipfs-webui/issues/1600 for consolidating repeat-use i18n keys, which I can probably take care of in the next week or two.

Thanks all (especially @jay-dee7!) for all of the work!

**Pull request merged**
